src/arith/solve_linear_equation.cc

```diff
@@ -0,0 +1,480 @@
+  // initialize V^{-1} x as x
+  for (size_t i = 0; i < num_vars; ++i) {
+    V.emplace_back(num_vars);
+    V.back()[i] = 1;
+    V_inv_x.push_back(system_to_solve->variables[i]);
+  }
+
+  // Transform formulas into rows of the matrix
+  // S_{mxn} V^{-1}_{nxn} x_{nx1} = U y, in which n is # of variables
+  // here we initialize S_{mxn} to be A, U to be identity matrix.
+  for (const PrimExpr& equation : system_to_solve->relations) {
+    if (const tir::EQNode* eq = equation.as<tir::EQNode>()) {
+      // a-b = sum_{i=0}^{n-1} variables[i] * coeff[i] + coeff[n]
+      Array<PrimExpr> coeffs = arith::DetectLinearEquation(
+          analyzer_problem.Simplify(eq->a - eq->b),
+          system_to_solve->variables);
+      if (!coeffs.empty()) {
+        std::vector<int64_t> row;
+        for (size_t j = 0; j < coeffs.size() - 1; ++j) {
+          PrimExpr c = coeffs[j];
+          if (const IntImmNode* ic = c.as<IntImmNode>()) {
+            row.push_back(ic->value);
+          } else {
+            // elements in matrix S V must be integers
+            // ignore equations that we cannot deal with.
+            LOG(WARNING) << "Cannot deal with non-integer coefficients, ignore equation "
+                         << equation;
+            row.clear();
+            break;
+          }
+        }
+
+        if (!row.empty()) {
+          // S V^{-1} (a-b) = Uy
+          // V is identity for now
+          S.push_back(row);
+          Uy.push_back(-coeffs[coeffs.size() - 1]);
+          continue;
+        }
+      }
+    }
+
+    // otherwise
+    rest.push_back(equation);
+  }
+
+  // After diagonalizing, we have
+  // S_{mxn} is the Smith normal form (diagonal matrix)
+  // V_{nxn} is invertible
+  // V_inv_x is V^{-1} \times x
+  // Uy is U \times y
+  SmithNormalFormDiag(&S, &V, &V_inv_x, &Uy);
+
+  Array<Var> new_vars;
+  Array<PrimExpr> new_relations;
+  Map<Var, PrimExpr> new_to_old_map;
+  Map<Var, PrimExpr> old_to_new_map;
+
+  // Simplify right hand sides
+  for (PrimExpr r : Uy) {
+    r = analyzer_problem.Simplify(r);
+  }
+
+  // Create the relations of the existence of a solution
+  for (size_t j = 0; j < S.size(); ++j) {
+    PrimExpr new_relation;
+    if (j >= num_vars || S[j][j] == 0) {
+      // The row of matrix is zero. A solution exists only if the Ub[j] is also zero
+      new_relation = (Uy[j] == 0);
+    } else {
+      // The diagonal element is non-zero. A solution exists only if the diagonal element
+      // is a divisor of the Ub[j]
+      new_relation = (floormod(Uy[j], std::abs(S[j][j])) == 0);
+    }
+    new_relation = analyzer_problem.Simplify(new_relation);
+    if (tir::is_const_int(new_relation, 0)) {
+      // unable to solve the system.
+      return IntConstraintsTransform(
+          system_to_solve,
+          IntConstraints(
+              /*variables=*/{},
+              /*ranges=*/{},
+              /*relations=*/{te::make_zero(DataType::Bool())}),
+          {}, {});
```

Comment:
I'm not sure, but I think src_to_dst still should map original variables to some dummy values (zero probably), this should prevent edge cases somewhere.

Reply:
In `check_solution` test I did see such dummy mapping helps to pass the check. while I'm not sure whether this is the right thing to do, since `src_to_dst` is normally defined to provide the "solution". How about we leave it empty here and revisit if later we find it is necessary?
For now in test I do this: https://github.com/apache/incubator-tvm/pull/5171/files#diff-b8ec9195578ec265834eddba2d5f5225R77-R79

Reply:
Ok